tidy3d/web/api/mode.py

```diff
@@ -137,6 +145,97 @@ def run(
+        Optional callback function called when downloading file with ``bytes_in_chunk`` as argument.
+
+
+    Returns
+    -------
+    List[ModeSolverData]
+        A list of ModeSolverData objects containing the results from each simulation in the batch. None is placed in the list for simulations that fail after all retries.
+    """
+    console = get_logging_console()
+
+    # Check type to make sure the user is submitting a list of mode solvers
+    if False in [type(x)== ModeSolver for x in mode_solvers]: 
+        console.print("[red]ERROR: Stopped running a batch of mode solver simulations. Please provide a list of mode solvers of type [cyan]`td.plugins.mode.mode_solver.ModeSolver` [red]to run a batch of mode solver simulations.")
+        return
+
+    # Result files where the data is stored after downloading
+    if results_files is None:
+        results_files = [f"{folder_name}/mode_solver_results_{i}.hdf5" for i in range(len(mode_solvers))]
+
+    results = [None] * len(mode_solvers)
+    
+    # Handling simulation to make sure there are more than one tries if simulation produces some error
+    def handle_simulation(index, retries=0):
+        try:
+            # We'll create a separate progress bar for batch simulations so we'll set verbose=False here for individual runs
+            result = run(mode_solvers[index], f"{task_name}_{index}", f"mode_solver_batch_{index}", folder_name, results_files[index], False, progress_callback_upload, progress_callback_download)
+            results[index] = result
+        except Exception as exc:
+            if retries < max_retries:
+                # Wait before retrying
+                time.sleep(retry_delay)  
+                handle_simulation(index, retries + 1)
+            else:
+                results[index] = None 
+
+    # Using multithreading to run mode solver simulations concurrently
+    with ThreadPoolExecutor(max_workers=max_workers) as executor:
+        futures = [executor.submit(handle_simulation, i) for i in range(len(mode_solvers))]
+        # Creating a progress bar for batch simulations using rich progress
+        if verbose:
+            with Progress(console=console) as progress:
+                pbar_message = f"[cyan]Running a batch of [deep_pink4]{len(mode_solvers)} mode solvers [cyan]in parallel.\n"
+                pbar = progress.add_task(pbar_message, total=len(mode_solvers))
+                for future in as_completed(futures):
+                    progress.advance(pbar)
+                    try:
+                        future.result()  # Add timeout if needed
+                    except TimeoutError:
+                        console.print(f"[red]A ModeSolver task has timed out.")
+
+                # Make sure the progress bar is complete
+                progress.update(pbar, completed=len(mode_solvers), refresh=True)
+                console.print("[green]Batch simulations completed successfully!")  
```

Comment:
batch of 'ModeSolver's maybe?

Also, how do we handle this if one of them actually failed? does the progress bar get completed still and do we still print this message?